### PR TITLE
feat(datasets): add experimental chromadb dataset

### PR DESCRIPTION
**Pull request by armandmasseaugit**
## Description

This PR adds a new experimental dataset `chromadb.ChromaDBDataset` to handle ChromaDB vector database collections, enabling seamless integration with AI/ML workflows that require vector storage and retrieval capabilities.

**Why ChromaDB?**
- ChromaDB is a popular open-source vector database for AI applications
- Essential for RAG (Retrieval-Augmented Generation) pipelines
- Enables semantic search and similarity matching in Kedro workflows
- Growing demand in the AI community for vector database integration

### What was implemented

✅ **Complete ChromaDBDataset implementation**
- Support for all ChromaDB client types: `ephemeral`, `persistent`, and `http`
- Full save/load functionality for documents, embeddings, and metadata
- Advanced querying capabilities with filters and search options
- Proper error handling and data validation
- Follows established Kedro dataset patterns

✅ **Comprehensive test suite**
- Unit tests covering all functionality and edge cases
- Tests for different ChromaDB client configurations
- Error handling validation
- Metadata and embeddings support testing

✅ **Complete documentation and configuration**
- API documentation with usage examples
- Integration with MkDocs navigation
- Proper dependency management in `pyproject.toml`
- Updated `RELEASE.md` with feature description

### Files changed

**Core implementation:**
- `kedro_datasets_experimental/chromadb/chromadb_dataset.py` - Main dataset implementation
- `kedro_datasets_experimental/chromadb/__init__.py` - Module exports

**Tests:**
- `kedro_datasets_experimental/tests/chromadb/test_chromadb_dataset.py` - Comprehensive test suite

**Documentation:**
- `docs/api/kedro_datasets_experimental/chromadb.ChromaDBDataset.md` - API documentation
- `mkdocs.yml` - Added to navigation under "Experimental LLM and AI"

**Configuration:**
- `pyproject.toml` - Added `chromadb>=0.4.0` dependency and test requirements
- `RELEASE.md` - Documented new experimental dataset

### Usage examples

```yaml
# YAML API - Persistent storage
my_documents:
  type: chromadb.ChromaDBDataset
  collection_name: "documents"
  client_type: "persistent"
  client_settings:
    path: "./vector_db"

# YAML API - HTTP client
remote_vectors:
  type: chromadb.ChromaDBDataset
  collection_name: "embeddings"
  client_type: "http"
  client_settings:
    host: "localhost"
    port: 8000
```

```python
# Python API
from kedro_datasets_experimental.chromadb import ChromaDBDataset

# Save documents with automatic embeddings
data = {
    "documents": ["AI is transforming industries", "Machine learning applications"],
    "ids": ["doc1", "doc2"],
    "metadatas": [{"category": "tech"}, {"category": "ml"}]
}

dataset = ChromaDBDataset(collection_name="knowledge_base")
dataset.save(data)

# Load and query documents
loaded = dataset.load()
print(loaded["documents"])
```

### Testing approach

```bash
# Install with ChromaDB support
pip install "kedro-datasets[chromadb]"

# Run the specific tests
pytest kedro_datasets_experimental/tests/chromadb/ -v
```

**Test coverage includes:**
- Basic save/load operations
- Different client types (ephemeral, persistent, HTTP config)
- Custom embeddings support
- Query filters and advanced loading
- Error handling for invalid data
- Metadata preservation
- Collection existence checking

### Integration with AI/ML workflows

This dataset enables powerful AI workflows in Kedro:
- **RAG pipelines**: Store and retrieve relevant documents for LLM context
- **Semantic search**: Find similar documents based on vector similarity  
- **Knowledge bases**: Build searchable document repositories
- **Embedding storage**: Persist vector representations for reuse
- **Multi-modal AI**: Store text, image, and other embeddings

## Checklist

- [x] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Updated `jsonschema/kedro-catalog-X.XX.json` if necessary
- [x] Added a description of this change in the relevant `RELEASE.md` file
- [x] Added tests to cover my changes
- [x] Received approvals from at least half of the TSC (required for adding a new, non-experimental dataset)

### CODE REVIEW

File: kedro-datasets/mkdocs.yml [hunk 1]
```diff
@@ -163,6 +163,7 @@ plugins:
           - api/kedro_datasets_experimental/index.md: Experimental features documentation
 
         Experimental LLM and AI:
+          - api/kedro_datasets_experimental/chromadb.ChromaDBDataset.md: ChromaDB vector database integration
```
Comment (ElenaKhaustova) [inline kedro-datasets/mkdocs.yml:172]: Could you please add it to the `nav` section, like it's done here: https://github.com/kedro-org/kedro-plugins/pull/1211/files#diff-15fdcc73534786fb1d6bdb24f0a3a2bf7af2d7e9385dffaa374594781b4ae233

File: kedro-datasets/kedro_datasets_experimental/tests/chromadb/test_chromadb_dataset.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+        # Verify the data
+        assert loaded_data["documents"] == sample_data["documents"]
+        assert loaded_data["ids"] == sample_data["ids"]
+        assert loaded_data["metadatas"] == sample_data["metadatas"]
+
+    def test_exists(self, chromadb_dataset, sample_data):
+        """Test the exists method."""
+        # Initially, collection shouldn't exist or be empty
+        assert not chromadb_dataset.exists()
```
Comment (ankatiyar) [inline kedro-datasets/kedro_datasets_experimental/tests/chromadb/test_chromadb_dataset.py:70]: This assertion fails when the whole test suite runs because the dataset has been saved in a previous test.

File: kedro-datasets/kedro_datasets_experimental/tests/chromadb/test_chromadb_dataset.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        """Test invalid client type raises error."""
+        with pytest.raises(DatasetError, match="Unsupported client_type: invalid"):
+            ChromaDBDataset(
+                collection_name="test",
+                client_type="invalid"
+            )
+
+    def test_http_client_config(self):
+        """Test HTTP client configuration."""
+        # This test creates the dataset but doesn't try to connect
```
Comment (ankatiyar) [inline kedro-datasets/kedro_datasets_experimental/tests/chromadb/test_chromadb_dataset.py]: When I run this test, I get this error -  `ValueError: Could not connect to a Chroma server. Are you sure it is running?`
The initialisation of the dataset shouldn't be needing to connect according to the comment, maybe something needs to be updated in the code?

File: kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
+"""``ChromaDBDataset`` loads and saves data from/to ChromaDB collections."""
+
+from __future__ import annotations
+
+import logging
+from pathlib import PurePosixPath
```
Comment (ElenaKhaustova) [inline kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py]: Some imports, including this one and below are not used anywhere

File: kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        """
+        self._collection_name = collection_name
+        self._client_type = client_type
+        self._client_settings = client_settings or {}
+        self._load_args = load_args or {}
+        self._save_args = save_args or {}
+        self.metadata = metadata
+
+        # Initialize ChromaDB client
+        self._client = self._create_client()
```
Comment (ElenaKhaustova) [inline kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py]: Can we explicitly annotate `_collection` and `_client` attributes to aid IDEs and type checkers?

File: kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+            # Add optional fields if present
+            if "metadatas" in data:
+                add_kwargs["metadatas"] = data["metadatas"]
+            if "embeddings" in data:
+                add_kwargs["embeddings"] = data["embeddings"]
+
+            # Add documents to collection
+            collection.add(**add_kwargs)
+
+        except Exception as e:
```
Comment (ElenaKhaustova) [inline kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py:250]:
```suggestion
        except Exception as e:
    raise DatasetError(
        f"Failed to load data from ChromaDB collection '{self._collection_name}'"
    ) from e
```

Using `from e` preserves the original traceback, which helps in Kedro logs.

File: kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+            raise DatasetError(f"Failed to save data to ChromaDB collection '{self._collection_name}': {e}")
+
+    def exists(self) -> bool:
```
Comment (ElenaKhaustova) [inline kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py:255]: Currently, `exists` creates a new collection as a side effect, which is probably not a desired behaviour.

```suggestion
    def exists(self) -> bool:
    """Check if the collection exists and contains any documents."""
    try:
        collection = self._client.get_collection(name=self._collection_name)
        return collection.count() > 0
    except Exception:
        return False

```

File: kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+
+import chromadb
+from kedro.io.core import (
+    AbstractDataset,
+    DatasetError,
+    get_filepath_str,
+    get_protocol_and_path,
+)
+
+logger = logging.getLogger(__name__)
```
Comment (ElenaKhaustova) [inline kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py]: Seems like you don't use logger as well.

File: kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+                f"Unsupported client_type: {self._client_type}. "
+                f"Must be one of: 'ephemeral', 'persistent', 'http'"
+            )
+
+    def _get_collection(self):
+        """Get or create the ChromaDB collection."""
+        if self._collection is None:
+            try:
+                self._collection = self._client.get_collection(name=self._collection_name)
+            except Exception:
```
Comment (ElenaKhaustova) [inline kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py]: Can we catch some specific exception here, like `NotFoundError` (if it exists for chromadb) rather than the broad one?

File: kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+            return chromadb.HttpClient(host=host, port=port, **{
+                k: v for k, v in self._client_settings.items() if k not in ["host", "port"]
+            })
+        else:
+            raise DatasetError(
+                f"Unsupported client_type: {self._client_type}. "
+                f"Must be one of: 'ephemeral', 'persistent', 'http'"
+            )
+
+    def _get_collection(self):
```
Comment (ElenaKhaustova) [inline kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py]: One improvement to consider: currently `_get_collection()` always creates a collection when it doesn’t exist, which introduces a side effect even during read operations (e.g., `exists()` or `load()`).

To make this behaviour explicit and more predictable, it might be worth adding a `create_if_missing` flag to `_get_collection()`.
That way, `save()` can safely auto-create collections, while `load()` and `exists()` can remain read-only.

Then you could call:

- `self._get_collection(create_if_missing=True)` in save()
- `self._get_collection(create_if_missing=False)` in load() / exists()

File: kedro-datasets/pyproject.toml [hunk 11]
```diff
@@ -232,6 +235,7 @@ test = [
     "accelerate<0.32",  # Temporary pin
     "adlfs~=2023.1",
     "biopython~=1.73",
+    "chromadb>=0.4.0",
```
Comment (ElenaKhaustova) [inline kedro-datasets/pyproject.toml]: Shall we bump the minimum `chromadb` version to >=1.0.0 (the latest stable release, the current latest is 1.2.2)?

**Review by ElenaKhaustova — COMMENTED** [review]
Great work on this implementation -  it’s looking really clean and solid!

I’ve left a few comments and small for refinement.

Happy to take another look and help with some manual testing once those updates are in.

**Comment by armandmasseaugit** [comment]
I get an error on matplotlib dataset, it comes from the main branch I guess

File: kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+            collection.add(**add_kwargs)
+
+        except Exception as e:
+            raise DatasetError(
+                f"Failed to save data to ChromaDB collection '{self._collection_name}': {e}"
+            ) from e
+
+    def exists(self) -> bool:
+        """Checks if the collection exists and contains data."""
+        try:
```
Comment (ElenaKhaustova) [inline kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py:257]: If `_get_collection()` returns `None` (collection doesn’t exist), `collection.count()` will raise an `AttributeError`.

```suggestion
        try:
            collection = self._collection or self._get_collection(create_if_missing=False)
            # In case both return None
            if collection is None:
                return False
            return collection.count() > 0
        except Exception:
            return False
```

File: kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+                f"Must be one of: 'ephemeral', 'persistent', 'http'"
+            )
+
+    def _get_client(self):
+        """Get or create the ChromaDB client."""
+        if self._client is None:
+            self._client = self._create_client()
+        return self._client
+
+    def _get_collection(self, create_if_missing: bool = True):
```
Comment (ElenaKhaustova) [inline kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py]: Can we please add a docstring line clarifying the `create_if_missing` return behavior?

File: kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+
+        # Prepare load arguments
+        load_args = {
+            "include": ["documents", "metadatas", "embeddings"],
+            **self._load_args
+        }
+
+        try:
+            # Use get() for retrieving all documents, query() requires search parameters
+            if "where" in load_args or "where_document" in load_args:
```
Comment (ElenaKhaustova) [inline kedro-datasets/kedro_datasets_experimental/chromadb/chromadb_dataset.py]: Should we also include `query_texts` here?

**Review by ElenaKhaustova — COMMENTED** [review]
Thank you - much improved from the previous iteration!

Left a couple of small suggestions (mainly an explicit None check in exists() and a doc tweak), but overall this looks good!
